### PR TITLE
feat: Add the stdlib_diff tool to compare gno and go standard libraries

### LINKED ISSUE #1310: Tool for comparing source code of standard libaries between Go and Gno
It would be useful to have a tool to do what is described in the title. Right now, when reviewing PRs, I'm comparing the source with Go's using [diffchecker](https://www.diffchecker.com/). I think we can create a simple tool that compares the directories and source code in GOROOT/src and those in GNOROOT/gnovm/stdlibs.

My idea:

- `misc/stdlibs_diff`
- Go executable for generating HTML files showing the diff.
- One HTML file per stdlib package.
- Show entire stdlibs/files we're missing, and entire stdlibs/files that have been added.
- Publish on https://gnolang.github.io/gno/stdlibs/

### PR DESCRIPTION
**Pull request by Villaquiranm**
continuing the work started on #1425 
thanks FloRichardAloeCorp for the great job on this issue 👍 
closes #1310 

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: misc/stdlib_diff/myers.go
Comment (Villaquiranm) [inline misc/stdlib_diff/myers.go:4]: > [thehowl](https://github.com/thehowl) [on Dec 13, 2023](https://github.com/gnolang/gno/pull/1425#discussion_r1425454096)
seeing as you are importing slices, please add a go.mod file in this directory putting the go version at go 1.21

fixed on eaf6a87b2bcf0dc7f61662ef23b9bc80cc1a26d6

File: misc/stdlib_diff/packagediff.go
Comment (Villaquiranm) [inline misc/stdlib_diff/packagediff.go:158]: >[thehowl](https://github.com/thehowl) [on Dec 13, 2023](https://github.com/gnolang/gno/pull/1425#discussion_r1425473328)
Prefer os.ReadDir

fixed on eaf6a87b2bcf0dc7f61662ef23b9bc80cc1a26d6

**Comment by Villaquiranm** [comment]
@thehowl I took a look of this, the current status:
1- Not done yet
2- I think it is already done, I can select the code and the + & - are not selected
3- I was thinking on implementing this using semaphores ? like that I could have the entire code on the same file as it is right now. If you think is necessary to use channels I could do that too.

4 and 5 are fixed on eaf6a87b2bcf0dc7f61662ef23b9bc80cc1a26d6
Only for 4 I think there are some folder we should avoid ? For exemple `cmd` & `vendor` as they have a lot of files and I think they are not related to standard lib ?


>Thank you for your contribution!
>
>Sorry for the wait, this is an initial to improve. Nice work, but you can improve.
>
>Some functionality missing on the generated report:
>
> 1- Formatting is still very funky. Use CSS to render tabs correctly; they should render as 4 spaces.

image
>
> 2- Ideally, for the added/deleted lines, make the + and - symbols non selectable. This makes it so that if I select the lines in either side of the page I can copy and paste the code without copying the symbols.
>
> 3- I believe the report's performance could be improved. Consider using the library I mention in the comments (diffmatchpatch) and parallelizing execution. You can parallelize execution by having a goroutine trying to first of all find all of the directories in the source and destination, and then sending off to a channel individual package names.
>
> 4- The tool does not currently perform diffing recursively. Ie. creating the report on my stdlibs it does not generate a diff for package regexp/syntax. Consider searching using something like filepath.WalkDir.
>
> 5- On the report, please make the lines which are == greyed out (putting css opacity: 0.75 probably works)

**Comment by thehowl** [comment]
@Villaquiranm 

1. Formatting: you're right, `+` / `-` are not selectable. Can you add line numbers as well, and align each column? (line number, + / - / no sign, code line), making sure that only the code is selectable?
2. For the diff'ing algo and performance, I actually found [this library](https://github.com/hexops/gotextdiff) when I was working on a side project some time ago. I believe it's the best for this use case. diffmatchpatch, which I linked in the original PR, is not ideal as it is for "character-by-character" diff (ie. what Google uses for google docs); not for the kind of diffs we use in programming. \
There are also likely some other "easy" performance wins. As you mentioned, we should skip directories which entirely don't exist in their Gno counterpart (in the index, we can just show `cmd` in red and say `missing in gno` or `missing in go` for libs like std). \
Parallelism may not be necessary; let's see what single-thread performance we can get, and if it's still slow after my suggestions use `pprof` to find out what's taking the most time. I think we can get <5s performance on most machines pretty easily even without it.
4. Can you make sure the index lists sub-directories as well?

**Comment by jefft0** [comment]
Removed the "review team" label because this is already reviewed by thehowl.

**Comment by Villaquiranm** [comment]
> @Villaquiranm
> 
> 1. Formatting: you're right, `+` / `-` are not selectable. Can you add line numbers as well, and align each column? (line number, + / - / no sign, code line), making sure that only the code is selectable?
> 2. For the diff'ing algo and performance, I actually found [this library](https://github.com/hexops/gotextdiff) when I was working on a side project some time ago. I believe it's the best for this use case. diffmatchpatch, which I linked in the original PR, is not ideal as it is for "character-by-character" diff (ie. what Google uses for google docs); not for the kind of diffs we use in programming. 
>    There are also likely some other "easy" performance wins. As you mentioned, we should skip directories which entirely don't exist in their Gno counterpart (in the index, we can just show `cmd` in red and say `missing in gno` or `missing in go` for libs like std). 
>    Parallelism may not be necessary; let's see what single-thread performance we can get, and if it's still slow after my suggestions use `pprof` to find out what's taking the most time. I think we can get <5s performance on most machines pretty easily even without it.
> 3. Can you make sure the index lists sub-directories as well?

Hey thanks a lot again for the review 👍 @thehowl .

For 1 I added the line numbers and they are not selectable

2- Thanks a lot for this library tip, I was actually taking a lot of time trying to figure this line diff algorithm but It was pretty straightforward with the one you recommended. The only downside is that the library just return the part of the file that have some changes, so to print the entire file I  needed to add some specific code to accomplish it.

3- subdirectories are listed on the index, if a library have subdirectories it will appear like a toggle list.
For performance I think we have something decent / usable the entire diff of the standard library takes around 2 seconds on my machine. 

I think there is still some improvements to do to the overall code, Just wanted to make sure this is the behaviour we'll like before doing some cleaning .

Thanks

**Comment by thehowl** [comment]
I suggest you not to rebase your PRs, but instead to merge in master; it makes my work when reviewing easier. And possibly yours as well, in managing conflicts. (There are none here I think, but this is good general guidance)

**Comment by thehowl** [comment]
How are you running the thing? I'm running with `go run . -src /usr/lib/go/src -dst ~/oc/gno/gnovm/stdlibs/ -out $(mktemp -d | tee /dev/stderr)` but it turns out an empty report.

**Comment by Villaquiranm** [comment]
> I suggest you not to rebase your PRs, but instead to merge in master; it makes my work when reviewing easier. And possibly yours as well, in managing conflicts. (There are none here I think, but this is good general guidance)

Sorry about that I'll keep it on mind

> How are you running the thing? I'm running with go run . -src /usr/lib/go/src -dst ~/oc/gno/gnovm/stdlibs/ -out $(mktemp -d | tee /dev/stderr) but it turns out an empty report.

I run it like this and for me it's working fine:
```sh
 go run . -src /opt/homebrew/Cellar/go/1.23.0/libexec/src -dst ../../gnovm/stdlibs -out ~/gnoreport/
2024/10/05 23:55:08 Building report...
2024/10/05 23:55:09 Report generation done!
```

**Comment by Villaquiranm** [comment]
@thehowl  
For some reason if I put 
`-dst ../../gnovm/stdlibs/` instead of `-dst ../../gnovm/stdlibs` is not working I'll check :(

**Comment by thehowl** [comment]
Okay, figured it out.

My `src` directory is a symlink. It works when I use its `realpath` (it's a unix command). Can you make sure symlinks are handled? (It's pretty common on linux).

The performance is satisfactory now. Thank you.

I have a bunch of comments to make on some of the looks. Mostly things about UX / confusing more than design wise, just to make sure this thing is decently usable. Tell me if you want 'em your way, or I can wait if you already have ideas.

- Can we make the toggles of differing files open by default?
- Can we make files that are missing in one or the other closed by default, but with a visually impactful note? (like red/bold ...)
- The split view doesn't align. Try to see it on github for what I mean (screenshot). Lines which are the same should be on the same level, ...

image

- Can we make sure to add also files and dirs that don't exist in go's stdlib? like `std`. This way, it's not just a compatibility document, but a general way for users to see what changes between the two standard libraries.
- Maybe it's best if you work your way from a Unified diff in the gotextdiff lib? I haven't examined your code yet, just spitballing based also on what I remember from using it.

**Comment by Villaquiranm** [comment]
@thehowl Thanks again for this new round of review :)
> My `src` directory is a symlink. It works when I use its `realpath` (it's a unix command). Can you make sure symlinks are handled? (It's pretty common on linux).
- handled symlinks
> * Can we make the toggles of differing files open by default?
- now the different files are open by default, but equal or one side missing files are not
> * Can we make files that are missing in one or the other closed by default, but with a visually impactful note? (like red/bold ...)
-  Missing files are  printed in colors Red if missing in destination and green if missing in source
> * The split view doesn't align. Try to see it on github for what I mean (screenshot). Lines which are the same should be on the same level, ...
I tried to align the best I could with the lib, There are still some lines that could be on the same line but in overall is better than before

> * Can we make sure to add also files and dirs that don't exist in go's stdlib? like `std`. This way, it's not just a compatibility document, but a general way for users to see what changes between the two standard libraries.
- I'm still missing this I'll work on it now
> * Maybe it's best if you work your way from a Unified diff in the gotextdiff lib? I haven't examined your code yet, just spitballing based also on what I remember from using it.
Yep I'm using unified diff but seems like it only keeps the line that have some differences, so entire parts of the file could be omited 
```go
edits := myers.ComputeEdits(span.URIFromPath(f.Src), f.srcContent, f.dstContent)
	unified := gotextdiff.ToUnified(f.Src, f.Dst, f.srcContent, edits)
```

Some sort of what github does(you can only see from line 8-12:
image

File: misc/stdlib_diff/templates/package_diff_template.html
Comment (thehowl) [inline misc/stdlib_diff/templates/package_diff_template.html:1]: Can you make the view take up 100% of the width?

image

(note it's not using all the available space on the right)

**Review by thehowl — APPROVED** [review]
There are some more improvements overall, but I think with these few changes we're good to merge and iterate. Thanks, and sorry for the delay in the review!

**Comment by thehowl** [comment]
btw, can you modify the gh-pages workflow? (.github/workflows/gh-pages)

so it creates the report automatically on each push. Maybe put the report in the subdir `stdlib_diff`, so it eventually shows up on https://gnolang.github.io/gno/stdlib_diff

I suggest you also try it out on a fork and link to a sample output, so we can check the workflow works correctly ahead of the merge.

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🔴 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (TERITORI:feat/stdlib_diff) is up to date with base (master): behind by 105 / ahead by 31

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

File: misc/stdlib_diff/README.md
Comment (Villaquiranm) [reply misc/stdlib_diff/README.md]: fixed on 2ff4f9e995834174744026cc5dfa57aa4e3573e7

File: misc/stdlib_diff/README.md
Comment (Villaquiranm) [reply misc/stdlib_diff/README.md]: fixed on 2ff4f9e995834174744026cc5dfa57aa4e3573e7

File: misc/stdlib_diff/README.md
Comment (Villaquiranm) [reply misc/stdlib_diff/README.md]: I removed the flag src_is_gno and I changed the way to obtain what each directory is suppose to contain gno or go files. on 2ff4f9e995834174744026cc5dfa57aa4e3573e7

File: misc/stdlib_diff/report.go
Comment (Villaquiranm) [reply misc/stdlib_diff/report.go]: changed to use filepath.EvalSymlinks on 2ff4f9e995834174744026cc5dfa57aa4e3573e7

File: misc/stdlib_diff/templates/package_diff_template.html
Comment (Villaquiranm) [reply misc/stdlib_diff/templates/package_diff_template.html:1]: fixed on 2ff4f9e995834174744026cc5dfa57aa4e3573e7

File: misc/stdlib_diff/filediff.go
Comment (Villaquiranm) [reply misc/stdlib_diff/filediff.go:195]: This is my way to convert the tabs to 4 spaces. is is working maybe is not the best way ?
Maybe I misunderstood
> 1- Formatting is still very funky. Use CSS to render tabs correctly; they should render as 4 spaces.

This line is doing the spaces to have the width equivalent of 4 spaces but when selected it is still a tab. I thought we had to convert it to real 4 spaces.
```css
 p {
            white-space-collapse: preserve;
            margin: 0;
            font-size: 1rem;
            font-weight: 400;
            tab-size: 4;
        }
```

**Comment by Villaquiranm** [comment]
> btw, can you modify the gh-pages workflow? (.github/workflows/gh-pages)
> 
> so it creates the report automatically on each push. Maybe put the report in the subdir `stdlib_diff`, so it eventually shows up on https://gnolang.github.io/gno/stdlib_diff
> 
> I suggest you also try it out on a fork and link to a sample output, so we can check the workflow works correctly ahead of the merge.

@thehowl  I followed your recommendations, here is the result :) thanks
https://villaquiranm.github.io/gno/stdlib_diff/

**Review by thehowl — APPROVED** [review]
https://github.com/TERITORI/gno/pull/24

**Comment by Villaquiranm** [comment]
@thehowl Merged your PR sorry for having this extra step :)
next time Will do it from my personal github account

**Comment by thehowl** [comment]
no worries :)

**Pull request merged**
